include/boost/graph/link_cut_trees.hpp

```diff
@@ -0,0 +1,280 @@
+//
+//=======================================================================
+// Copyright 2019
+// Author: Yi Ji
+//
+// Distributed under the Boost Software License, Version 1.0. (See
+// accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//=======================================================================
+//
+#ifndef BOOST_LINK_CUT_TREES_HPP
+#define BOOST_LINK_CUT_TREES_HPP
+
+#include <boost/graph/properties.hpp>
+#include <boost/unordered_map.hpp>
+
+namespace boost
+{
+    template<class ElementParentMap, class ElementChildMap = ElementParentMap>
+    class link_cut_trees
+    {
+    public:
+        inline link_cut_trees(ElementParentMap p, ElementChildMap l, ElementChildMap r) : parent(p), left(l), right(r) {}
+        
+        inline link_cut_trees(const link_cut_trees<ElementParentMap, ElementChildMap> &c)
+            : parent(c.parent), left(c.left), right(c.right) {}
+
+        template <class Element>
+        inline void make_tree(Element x)
+        {
+            make_path(x);
+        }
+
+        template <class Element>
+        inline Element find_root(Element x)
+        {
+            return find_tail(expose(x));
+        }
+
+        template <class Element>
+        inline void link(Element x, Element y) // Element x must be a tree root
+        {
```

Comment:
Can you assert that x is a tree root so that making that mistake is caught during debugging? Maybe it's too difficult to test, I haven't actually looked into it yet.

Reply:
But looking at it again, maybe:
`BOOST_ASSERT(find_root(x) == x);`

Reply:
The amortized time complexity is _O(log n)_ for both `link` and `find_root`; However, if `find_root` is called inside each `link` call, then I'm afraid the amortized time complexity might not remain the same.

Reply:
That's true, it will make the time complexity worse. In my opinion, that's OK in Debug mode, where a developer is concerned with correctness, not performance.
I definitely think this is a nice safeguard to have, but it's not critical.

Let's ask @jzmaddock what he thinks.

Reply:
@jeremy-murphy I've just realized that `BOOST_ASSERT` can be optinally disabled using certain [macros](https://www.boost.org/doc/libs/1_67_0/libs/assert/doc/html/assert.html#boost_assert). So I've added the assertion as suggested :) If users would like to skip them for best performance, they can easily achieve that.

Reply:
Ah, OK, cool, that way everyone is happy.